doozerlib/assembly.py

```diff
@@ -0,0 +1,138 @@
+        if type(c) is not dict:
+            return a
+        for k, v in a.items():
+            if k.endswith('!'):  # full dominant key
+                k = k[:-1]
+                c[k] = v
+            elif k.endswith('?'):  # default value key
+                k = k[:-1]
+                if k not in c:
+                    c[k] = v
+            else:
+                if k in c:
+                    c[k] = merger(a[k], c[k])
+                else:
+                    c[k] = a[k]
+        return c
+
+    raise TypeError(f'Unexpected value type: {type(a)}: {a}')
+
+
+def assembly_group_config(releases_config: Model, assembly: str, group_config: Model) -> Model:
+    """
+    Returns a group config based on the assembly information
+    and the input group config.
+    :param releases_config: A Model for releases.yaml.
+    :param assembly: The name of the assembly
+    :param group_config: The group config to merge into a new group config (original Model will not be altered)
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return group_config
+
+    target_assembly = releases_config.releases[assembly].assembly
+
+    if target_assembly.basis.assembly:  # Does this assembly inherit from another?
+        # Recursive apply ancestor assemblies
+        group_config = assembly_group_config(releases_config, target_assembly.basis.assembly, group_config)
+
+    target_assembly_group = target_assembly.group
+    if not target_assembly_group:
+        return group_config
+
+    return Model(dict_to_model=merger(target_assembly_group.primitive(), group_config.primitive()))
+
+
+def assembly_metadata_config(releases_config: Model, assembly: str, meta_type: str, distgit_key: str, meta_config: Model) -> Model:
+    """
+    Returns a group config based on the assembly information
+    and the input group config.
+    :param releases_config: A Model for releases.yaml.
+    :param assembly: The name of the assembly
+    :param meta_type: 'rpm' or 'image'
+    :param distgit_key: The name of the underlying component
+    :param meta_config: The meta's config object
+    :return: Returns a computed config for the metadata (e.g. value for meta.config).
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return meta_config
+
+    target_assembly = releases_config.releases[assembly].assembly
+
+    if target_assembly.basis.assembly:  # Does this assembly inherit from another?
+        # Recursive apply ancestor assemblies
+        meta_config = assembly_metadata_config(releases_config, target_assembly.basis.assembly, meta_type, distgit_key, meta_config)
+
+    config_dict = meta_config.primitive()
+
+    component_list = target_assembly.members[f'{meta_type}s']
+    for component_entry in component_list:
+        if component_entry.distgit_key == '*' or component_entry.distgit_key == distgit_key:
+            config_dict = merger(component_entry.metadata.primitive(), config_dict)
```

Comment:
The result of `merger` depends on the order of `component_entry` in the list. It seems to me we should merge component_entry for a specific distgit_key on the top of the wirdcard one in case of overriding.

Reply:
We might not ultimately need the mechanic, but I did intend for order to matter -- anticipating more exotic matching at some point. If it does prove to be overkill / confusing in the future, we can convert this to a dict and eliminate the expectation of ordering.

Reply:
ack